jep-014a-string-functions.md

```diff
@@ -0,0 +1,296 @@
+# String Functions
+
+|||
+|---|---
+| **JEP**    | 14a
+| **Author** | Maxime Labelle, Chris Armstrong (GorillaStack), Richard Gibson
+| **SemVer** | MINOR
+| **Status**| draft
+| **Created**| 13-October-2022
+| **Obsoletes**| [JEP-14](./jep-014-string-functions.md)
+
+## Addendum
+
+|Date|Description
+|---|---|
+|15-March-2023|Clarified error type precedence.
+
+## Abstract
+
+This JEP introduces a core set of useful string manipulation functions. Those functions are modeled from functions found in popular programming languages such as JavaScript and Python.
+
+## Specification
+
+Some string manipulation functions bring the new concept of _optional arguments_ to JMESPath functions. The specification paragraph on function evaluation must thus be changed accordingly – highlighted in **bold** in the text below:
+
+_Functions can ~~either~~ have a specific arity, **a range of valid – minimum and maximum – number of arguments** or be variadic with a minimum number of arguments. If a function-expression is encountered where the arity does not match or the minimum number of arguments for a variadic function is not provided, then implementations must indicate to the caller that an invalid-arity error occurred. How and when this error is raised is implementation specific._
+
+Some functions accept number arguments which are further constrained to integers or even non-negative integers. This JEP specifies a new error
+type `invalid-value` by updating the paragraph on type constraints from the specification like so:
+
+_Each function signature declares the types of its input parameters. If any type constraints are not met, implementations must indicate that an `invalid-type` error occurred. **If a function parameter accepts values constrained to a specific subset of a type and those constraints are not met, implementations must report that an `invalid-value` error occurred.**_
+
+_The [initial version of this JEP](./jep-014-string-functions.md) had a provision stating that_ “How and when those errors are raised is implementation specific”. _This provision has been removed. Implementation must perform type-checking for all function parameters_ before _attempting to evaluate the set of valid values for a given type._
```

Comment:
The requirement of type validation before range validation should not be limited to string functions... I think this deserves its own JEP rather than a replacement of JEP-14 (in which case I don't think it will be necessary to touch JEP-14 at all, because "_How and when those errors are raised is implementation specific_" can be interpreted within the context of that constraint (e.g., "_how_" might be "by exception" vs. "in a separate return value" and "_when_" might be "when parsing" [where possible] vs. "when evaluating").